core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -64,672 +83,1650 @@
             }
-            childStages = createChildStages.apply(bucketToPartition);
-        }
 
-        stage.addStateChangeListener(newState -> {
-            if (newState == FLUSHING || newState.isDone()) {
-                childStages.forEach(SqlStageExecution::cancel);
+            if (state == DistributedStagesSchedulerState.FAILED) {
+                StageFailureInfo stageFailureInfo = distributedStagesScheduler.getFailureCause()
+                        .orElseGet(() -> new StageFailureInfo(toFailure(new VerifyException("distributedStagesScheduler failed but failure cause is not present")), Optional.empty()));
+                ErrorCode errorCode = stageFailureInfo.getFailureInfo().getErrorCode();
+                if (shouldRetry(errorCode)) {
+                    long delayInMillis = min(retryInitialDelay.toMillis() * ((long) pow(2, currentAttempt.get())), retryMaxDelay.toMillis());
+                    currentAttempt.incrementAndGet();
+                    scheduleRetryWithDelay(delayInMillis);
+                }
+                else {
+                    stageManager.getDistributedStagesInTopologicalOrder().forEach(stage -> {
+                        if (stageFailureInfo.getFailedStageId().isPresent() && stageFailureInfo.getFailedStageId().get().equals(stage.getStageId())) {
+                            stage.fail(stageFailureInfo.getFailureInfo().toException());
+                        }
+                        else {
+                            stage.abort();
+                        }
+                    });
+                    queryStateMachine.transitionToFailed(stageFailureInfo.getFailureInfo().toException());
+                }
             }
         });
+        return Optional.of(distributedStagesScheduler);
+    }
 
-        stageLinkages.put(stageId, new StageLinkage(plan.getFragment().getId(), parent, childStages));
+    private boolean shouldRetry(ErrorCode errorCode)
+    {
+        return retryPolicy == RetryPolicy.QUERY && currentAttempt.get() < maxRetryAttempts && (errorCode == null || errorCode.getType() != USER_ERROR);
+    }
 
-        return stages.build();
+    private void scheduleRetryWithDelay(long delayInMillis)
+    {
+        try {
+            schedulerExecutor.schedule(this::scheduleRetry, delayInMillis, MILLISECONDS);
+        }
+        catch (Throwable t) {
+            queryStateMachine.transitionToFailed(t);
+        }
     }
 
-    public BasicStageStats getBasicStageStats()
+    private synchronized void scheduleRetry()
     {
-        List<BasicStageStats> stageStats = stages.values().stream()
-                .map(SqlStageExecution::getBasicStageStats)
-                .collect(toImmutableList());
+        try {
+            checkState(distributedStagesSchedulingTask != null, "schedulingTask is expected to be set");
+
+            // give current scheduler some time to terminate, usually it is expected to be done right away
+            distributedStagesSchedulingTask.get(5, MINUTES);
```

Comment:
What if we don't give it time to terminate? What if it takes longer than 5 minutes?

Reply:
> What if we don't give it time to terminate?

In an event of a failure all `PipelinedStageExecution`'s will be aborted. Once all `PipelinedStageExecution`s are aborted the `PipelinedDistributedStagesScheduler#schedule` method is expected to return (since all stage executions are either failed or aborted there's nothing else to be scheduled). 

However this is inherently racy, as failure handling and scheduling happen in different threads. Thus the "schedule" method is given "some" time to terminate. Normally it shouldn't take more than a couple of milliseconds / microseconds.

> What if it takes longer than 5 minutes?

In normal circumstances it should never happen unless there's a bug. However if there is a bug it is better to fail a query than to wait indefinitely (or to have a thread leak).

The timeout is set to a high value to avoid triggering a timeout accidentally (e.g.: due to a bad OS scheduling). 5 minutes is probably way too high, but since it should never happen in practice I thought it is better to be on a safer side :-)